Fixes 2932: let the capi version to be discovered

enxebre opened the pull request:
Add the ability to override CAPI group via env variable and discovers the API version.

This change adds detection for an environment variable to specify the group for the clusterapi resources. If the environment variable `CAPI_GROUP` is specified, then it will be used instead of the default. This also decouples the API group from the version and let the latter to be discovered dynamically.

Fixes https://github.com/kubernetes/autoscaler/issues/2932

enxebre commented:
/hold
To do some manual validations
cc @frobware @elmiko

enxebre commented:
/hold cancel

elmiko commented:
> Codewise looks good, just wondering whether we want to be calling all the variables etc Machine API or Cluster API, do we have a preference?

agreed, i think if we can use "cluster api" here it will reduce any confusion for someone reading this. at the very least we should name the external bits (eg env variable) to reflect cluster api.

frobware commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -281,15 +296,16 @@ func newMachineController(
-	machineSetResource, _ := schema.ParseResourceArg(fmt.Sprintf("machinesets.%v", defaultMachineAPI))
+	machineSetResource, _ := schema.ParseResourceArg(fmt.Sprintf("machinesets.%v", specifiedMachineAPI))
 	if machineSetResource == nil {
 		panic("MachineSetResource")
```
This should really return an error.

enxebre replied:
mm why? atm `MachineDeployment`, `MachineSets and `Machine` are intentionally assumed to exist so it seems reasonable to me to panic here?

frobware replied:
Because it makes unit tests easier to write and assert for expected behaviour.

enxebre replied:
ok make sense, let's address this separately from this PR. I'll move the capi resource expectations into its own unit tested `func`

frobware commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:
```diff
@@ -991,3 +992,110 @@ func TestControllerMachineSetNodeNamesUsingStatusNodeRefName(t *testing.T) {
+
+func TestControllerGetAPIVersionGroup(t *testing.T) {
+	expected := "myversion.mygroup"
+	err := os.Setenv(machineAPIVersionGroupEnvVar, expected)
```
nit: `if err := os.Setenv(machineAPIVersionGroupEnvVar, expected); err != nil` ...

frobware commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:
```diff
@@ -991,3 +992,110 @@ func TestControllerMachineSetNodeNamesUsingStatusNodeRefName(t *testing.T) {
+		t.Fatalf("Incorrect number of MachineDeployments, expected 1, got %d", l)
+	}
+
+	machineSets, err := controller.listMachineSets(testNamespace, labels.Everything())
```
Is this belt and braces? We're testing Machine Deployments in this test; are just verifying that it doesn't error? If so could we collapse `TestControllerGetAPIVersionGroupWithMachineSets` and `TestControllerGetAPIVersionGroupWithMachineDeployments`.

enxebre replied:
`TestControllerGetAPIVersionGroupWithMachineSets` seems pretty much covered by `TestControllerGetAPIVersionGroupWithMachineDeployments` may be just drop the first one and rename the latter one to `TestControllerGetAPIVersionGroup`, cc @elmiko ?

elmiko replied:
that sounds fair to me, i originally added both tests since it wasn't clear to me if there were different mechanisms for machinesets and machinedeployments.

enxebre commented:
>Codewise looks good, just wondering whether we want to be calling all the variables etc Machine API or Cluster API, do we have a preference?

>agreed, i think if we can use "cluster api" here it will reduce any confusion for someone reading this. at the very least we should name the external bits (eg env variable) to reflect cluster api.

I renamed to CAPI consistently and move code to fetch the preferred version into unit tested `getAPIGroupPreferredVersion`

enxebre commented:
>in general it looks good to me, deduplicating the tests seems like a good idea.
 
Addressed comments and meld into one commit. PTAL @elmiko @frobware @JoelSpeed

frobware commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:
```diff
@@ -76,7 +79,19 @@ func mustCreateTestController(t *testing.T, testConfigs ...*testConfig) (*machin
+		Fake: &clientgotesting.Fake{
+			Resources: []*v1.APIResourceList{
+				{
+					GroupVersion: "machine.openshift.io/v1beta1",
```
Why do we reference openshift here?

frobware replied:
OK, I see. We're testing discovery. But, I would still choose a more agnostic name for the case that should not be found.

enxebre replied:
this is used for validating env variable overridability, let me update it to something like "custom" to avoid confusion.

enxebre replied:
Updated to reference `customCAPIGroup = "custom.x-k8s.io"`

frobware commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -332,6 +357,28 @@ func newMachineController(
+		if group.Name == APIGroup {
+			foundAPIGroup = true
+			APIVersion = group.PreferredVersion.Version
+			break
```
Why not just `return` at this point? You can drop some variables and conditionals below.

frobware commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:
```diff
@@ -136,7 +153,7 @@ func createTestConfigs(specs ...testSpec) []*testConfig {
 		config.machineSet = &MachineSet{
 			TypeMeta: v1.TypeMeta{
-				APIVersion: "cluster.x-k8s.io/v1alpha2",
+				APIVersion: "cluster.x-k8s.io/v1alpha3",
```
Can these strings literals not use `defaultCAPIGroup`?

enxebre commented:
>The issue (#2934) talks about discovering the version but, unless I'm mistaken, this only allows the group to be overridden. True?

This enable an env variable to override the default CAPI group and makes the discovers the API version (https://github.com/kubernetes/autoscaler/issues/2932) see https://github.com/kubernetes/autoscaler/pull/2934/files#diff-89363471d762322b93c9bcfec57b5af8R302. Let me update the PR desc with the commit message.

frobware commented:
/approve

enxebre commented:
/area provider/cluster-api

elmiko commented:
/lgtm
